### PR TITLE
Implement mechanism for ensuring that all extensions share an AMP version with the main binary.

### PR DESCRIPTION
**Pull request by cramforce**
For each extension this checks the version, and if it is the wrong version, will ignore the JS file and initiate a load of the correct version.

- experiment guarded. This is likely to break some stuff.
- based on compiled-in version only. rtv is ignored, because we only care about code here, since config is only given in main binary.

Implements locking part of #7238

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -440,6 +440,7 @@ function buildExtensionJs(path, name, version, options) {
     // See https://github.com/ampproject/amphtml/issues/3977
     wrapper: options.noWrapper ? '' : ('(self.AMP=self.AMP||[])' +
         '.push({n:"' + name + '",' + priority +
+        'v:"' + internalRuntimeVersion + '",' +
```
Comment (dvoytenko) [inline gulpfile.js]: Does this include RTV prefix? I always get confused where it is added and where it's not. Does it matter?

File: src/runtime.js [hunk 2]
```diff
@@ -889,3 +895,48 @@ export function registerElementForTesting(win, elementName) {
...
+    return false;
+  }
+  if (typeof fnOrStruct == 'function') {
+    return false;
+  }
+  const version = fnOrStruct.v;
+  // This is non-obvious, but we only care about the release version,
+  // not about the full rtv version, because these only differ
+  // in the config that is fully determined by the primary binary.
+  if ('$internalRuntimeVersion$' == version) {
```
Comment (dvoytenko) [inline src/runtime.js]: Ditto...

File: src/runtime.js [hunk 3]
```diff
@@ -889,3 +895,48 @@ export function registerElementForTesting(win, elementName) {
...
+          `[custom-element="${fnOrStruct.n}"]:not([i-amphtml-inserted])`);
+  dev().assert(scriptInHead, 'Expected to find script for extension: %s',
+      fnOrStruct.n);
+  if (!scriptInHead) {
+    return false;
+  }
+  // Mark the element as being replace, so that the loadExtension code
+  // assumes it as not-present.
+  scriptInHead.removeAttribute('custom-element');
+  scriptInHead.setAttribute('i-amphtml-replaced-with-version', fnOrStruct.n);
```
Comment (dvoytenko) [inline src/runtime.js]: Should it be `fnOrStruct.v`?

File: src/runtime.js [hunk 4]
```diff
@@ -889,3 +895,48 @@ export function registerElementForTesting(win, elementName) {
...
+ *
+ * @param {!Window} win
+ * @param {function(!Object)|ExtensionPayloadDef} fnOrStruct
+ * @return {boolean}
+ */
+function maybeLoadCorrectVersion(win, fnOrStruct) {
+  if (!isExperimentOn(win, 'version-locking')) {
+    return false;
+  }
+  if (typeof fnOrStruct == 'function') {
```
Comment (jridgewell) [inline src/runtime.js]: Which extensions still use a function?

File: src/runtime.js [hunk 5]
```diff
@@ -889,3 +895,48 @@ export function registerElementForTesting(win, elementName) {
...
+  // not about the full rtv version, because these only differ
+  // in the config that is fully determined by the primary binary.
+  if ('$internalRuntimeVersion$' == version) {
+    return false;
+  }
+  const scriptInHead = win.document.head.querySelector(
+          `[custom-element="${fnOrStruct.n}"]:not([i-amphtml-inserted])`);
+  dev().assert(scriptInHead, 'Expected to find script for extension: %s',
+      fnOrStruct.n);
+  if (!scriptInHead) {
```
Comment (jridgewell) [inline src/runtime.js]: Can you explain this one? Why wouldn't we insert one even if it didn't exist?

File: src/runtime.js (hunk 3 above)
Comment (jridgewell) [inline src/runtime.js]: `.v`?

File: src/runtime.js (hunk 4 above)
Comment (dvoytenko) [inline src/runtime.js]: Bunch of code that wants AMP to initialize does, not necessarily extensions. E.g. viewer integration scripts (e.g. from Google Search) do that and bunch of PWA functions.

File: gulpfile.js (hunk 1 above)
Comment (cramforce) [inline gulpfile.js]: I have a comment at the use site for this. We care about the version of the code here, not about the version of the config.

File: src/runtime.js (hunk 4 above)
Comment (cramforce) [inline src/runtime.js]: Basically non-AMP code.

File: src/runtime.js (hunk 2 above)
Comment (cramforce) [inline src/runtime.js]: Existing comment OK?

File: src/runtime.js (hunk 5 above)
Comment (cramforce) [inline src/runtime.js]: There is an assertion above that makes this code unreachable in development. I think in prod it makes sense to just move forward as if nothing happened in this case.

File: src/runtime.js (hunk 3 above)
Comment (cramforce) [inline src/runtime.js]: Sorry, tried a new name for the attribute :)

I want the name here for easier traceability of the DOM.

File: src/runtime.js (hunk 2 above)
Comment (dvoytenko) [inline src/runtime.js]: Yes.

File: src/runtime.js (hunk 3 above)
Comment (dvoytenko) [inline src/runtime.js]: Ping on this. Otherwise LGTM.

File: src/runtime.js [hunk 6]
```diff
@@ -889,3 +895,50 @@ export function registerElementForTesting(win, elementName) {
...
+ * This is currently guarded by the 'version-locking' experiment.
+ * With this active, all scripts in a given page are guaranteed
+ * to have the same AMP release version.
+ *
+ * @param {!Window} win
+ * @param {function(!Object)|ExtensionPayloadDef} fnOrStruct
+ * @return {boolean}
+ */
+function maybeLoadCorrectVersion(win, fnOrStruct) {
+  if (!isExperimentOn(win, 'version-locking')) {
```
Comment (dvoytenko) [inline src/runtime.js]: Do you want to do instead `isExperimentOnAllowUrlOverride`?

Comment (cramforce) [inline src/runtime.js]: Could you explain why you think that makes sense?

**Pull request merged**